File: src/main/java/com/github/albertosh/flitetrakr/sync/services/ConnectionService.java
package com.github.albertosh.flitetrakr.sync.services;

import com.github.albertosh.flitetrakr.model.Connection;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Basic in memory storage
 */
public class ConnectionService implements IConnectionService {

    private final Set<String> cities;
    private final Map<String, List<Connection>> fromConnections;

    public ConnectionService() {
        fromConnections = new HashMap<>();
        cities = new HashSet<>();
    }

    @Override
    public void addConnection(Connection connection) {
        List<Connection> cityConnections = fromConnections.get(connection.getFrom());
        if (cityConnections == null) {
            cityConnections = new ArrayList<>();
            fromConnections.put(connection.getFrom(), cityConnections);
        }
        cityConnections.add(connection);
        cities.add(connection.getFrom());
        cities.add(connection.getTo());
    }

    @Override
    public Optional<Connection> recoverConnection(String from, String to) {
        List<Connection> cityConnections = fromConnections.get(from);
        if (cityConnections == null) {
            return Optional.empty();
        } else {
            for (Connection c : cityConnections) {
                if (c.getTo().equals(to))
                    return Optional.of(c);
            }
            return Optional.empty();
        }
    }

    @Override
    public Set<String> getCities() {
        return new HashSet<>(cities);
    }

    @Override
    public Set<String> getDestiniesFromCity(String city) {
        List<Connection> cityConnections = fromConnections.get(city);
        Set<String> result = new HashSet<>();
        if (cityConnections == null) {
            return result;
        } else {
            cityConnections.stream()
                    .map(Connection::getTo)
                    .forEach(result::add);
            return result;
        }
    }
}

File: flitetrakr_reactive_kotlin/src/main/kotlin/com/github/albertosh/flitetrakr/services/ConnectionService.kt
package com.github.albertosh.flitetrakr.services

import com.github.albertosh.flitetrakr.model.Connection
import com.github.albertosh.flitetrakr.util.language.LanguageUtils
import com.github.albertosh.flitetrakr.util.language.Message
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single

interface IConnectionService {

    fun addConnection(connection: Connection): Completable

    fun recoverConnection(from: String, to: String): Single<Connection>

    fun getCities(): Flowable<String>

    fun getDestiniesFromCity(city: String): Flowable<String>

}

sealed class ConnectionServiceError(message: String) : RuntimeException(message) {
    object connectionNotFound : ConnectionServiceError(
            LanguageUtils.getMessage(Message.CONNECTION_NOT_FOUND))
}

/**
 * Basic in memory storage
 */
class ConnectionService : IConnectionService {

    private val storedCities = mutableSetOf<String>()
    private val fromConnections = mutableMapOf<String, MutableList<Connection>>()

    override fun getCities(): Flowable<String> {
        return Flowable.fromIterable(storedCities)
    }

    override fun addConnection(connection: Connection): Completable {
        return Completable.create {
            var cityConnections = fromConnections.get(connection.from)
            if (cityConnections == null) {
                cityConnections = mutableListOf<Connection>()
                fromConnections.put(connection.from, cityConnections)
            }
            cityConnections.add(connection)
            storedCities.add(connection.from)
            storedCities.add(connection.to)
        }
    }

    override fun recoverConnection(from: String, to: String): Single<Connection> {
        return Single.create { source ->
            val connection = fromConnections.get(from)
                    ?.firstOrNull { it.to == to }
            if (connection != null)
                source.onSuccess(connection)
            else
                source.onError(ConnectionServiceError.connectionNotFound)
        }
    }

    override fun getDestiniesFromCity(city: String): Flowable<String> {
        return Flowable.fromIterable(
                fromConnections.get(city)
                        ?.map { it.to }
                        ?.toSet()
                        ?: emptySet()
        )
    }
}